src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -130,47 +142,104 @@ public AWSSecretsManagerPlugin(
 
   @Override
   public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
-    final Pair<String, Region> secretKey = Pair.of(secretId, region);
+
     final Properties properties = new Properties();
     properties.putAll(connectionUrl.getOriginalProperties());
-    Secret secret = SECRET_CACHE.get(secretKey);
+
+    boolean secretWasFetched = updateSecret(false);
 
     try {
-      // Attempt to open initial connection with cached secret.
-      attemptConnectionWithSecrets(properties, secret, connectionUrl);
-
-    } catch (SQLException connectionFailedException) {
-      // Rethrow the exception unless it was because user access was denied.
-      // In that case, retry with new credentials.
-      if (!SQLSTATE_ACCESS_ERROR.equals(connectionFailedException.getSQLState())) {
-        throw connectionFailedException;
-
-      } else {
-        try {
-          secret = getCurrentCredentials();
-
-        } catch (SecretsManagerException | JsonProcessingException getSecretsFailedException) {
-          this.logger.logError(ERROR_GET_SECRETS_FAILED);
-          throw new SQLException(ERROR_GET_SECRETS_FAILED);
+      applySecretToProperties(properties);
+      attemptToLogin(properties, connectionUrl);
+
+    } catch (SQLException exception) {
+      if (isLoginUnsuccessful(exception) && !secretWasFetched) {
+        // Login unsuccessful with cached credentials
+        // Try to re-fetch credentials and try again
+
+        secretWasFetched = updateSecret(true);
```

Comment:
at this point if you get the secret here you can go ahead

Reply:
If secretWasFetched  is false then the credentials hasn't been updated and it makes no sense to try to login again so we just throw original exception.